runtime/altair/src/lib.rs

```diff
@@ -1350,6 +1352,38 @@ impl pallet_membership::Config for Runtime {
 	type WeightInfo = pallet_membership::weights::SubstrateWeight<Self>;
 }
 
+parameter_types! {
+	#[derive(Clone, PartialEq, Eq, Debug, TypeInfo, Encode, Decode, MaxEncodedLen)]
+	pub const MaxFeedersPerKey: u32 = 10;
+	pub const FirstValueFee: Fee = Fee::Balance(10 * CFG); // TODO: Specify
+}
+
+impl pallet_oracle_feed::Config for Runtime {
+	type FirstValuePayFee = FeeToTreasury<Fees, FirstValueFee>;
+	type OracleKey = OracleKey;
+	type OracleValue = Quantity;
```

Comment:
Is it really `Quantity`? I would have imagined `Rate` for loans or `Ratio` for the orderbook

Reply:
`Rate` sure not because it has 27 decimals with a few integer parts. It could be a `Ratio` which is exactly the same time as `Quantity` (both alias the same exact type). We should probably unify these types to avoid confusion.

Regarding the name, I think you're right, and `Ratio` fits better to specify a "price"

Reply:
@lemunozm can you make this an enum already? I think we will have non ratio types in the future. WDYT?

Reply:
Make sense to avoid future migrations

Reply:
@mustermeiszer thinking about this. Do you think it would be interesting to have different pallet instances for `pallet-oracle-feed` in case of different values instead?

I say this to avoid, for example, a feeder feeding with a Key as `Isin` and a Value as a non-radio type, mixing keys with non related values.

Reply:
I think it all depends on how the consumer wants to consume. No strong opinion on it

Reply:
I think instantiating the `pallet-oracle-feed` makes more sense for the mentioned reason. That can be done in a follow up PR. But @mustermeiszer should have the stronger say here.

Reply:
Agree!

Once this last point is solved, I think we can merge this PR.

Reply:
After another thought, definitely I think it's better to use different pallet instances to maintain always well-formed key-value correlations on-chain. If some consumer needs different value types, we can always compound them into an enum using a runtime type that implements `ValueProvider` in the runtime. Similar to what we already do with `OracleConverterBridge`.

Reply:
I am really undecided here. 

PRO: 
* strong typed relationship
* cleaner from a technical perspective

CON: 
* We have 70 pallets right now, the limit is 255 - not really close, but we are already having more pallets than most parachains IIRC
* External integrations - It is just harder to look at the right place but I might over-value that

I have even thought about allowing this API

```
		pub fn feed(
lemunozm marked this conversation as resolved.
			origin: OriginFor<T>,
			key: T::OracleKey,
			value: BoundedVec<T::MaxScaleSizeValue>,
		) -> DispatchResultWithPostInfo {
```

and upon fetching/aggregating we decode into the needed type and fail if it is not correct. That would be the most flexible.

For example with currently we have needs for
* `ISIN` -> `Price`
* `(Currency, Currency)` -> `Price`

Should we really have an instance of this pallet for each combination that we need on-chain? There will be more, I can promise that ^^

Reply:
🤔

> Should we really have an instance of this pallet for each combination that we need on-chain?

I think both relations above can be in the same pallet instance. The pallet instance can have the Key as an enum but the Value as `Price`. That way, any Key enum variant will map to a correct Value type.

The amount of instances will depend only on the amount of different kinds of values, not keys. Mostly, I can see `Price,` `Balance`, and maybe some id or similar (?).

WDYT?

Reply:
The encoding is always ensured when fetching the keys. Basically anywas what we are already doing when accceotinbg an extrinsic, but just during a runtime call. It is then an invalid key. 

But lets go with instances if you think it is better. Should we already make the current `key` then an enum?

Reply:
> Should we already make the current key then an enum?

It currently is!